### PR TITLE
Refactored Framework and Added 3D Visualization Capabilities

### PR DESCRIPTION
**Pull request by MattsonCam**
This pr refactors the segmentation framework, updated the uv environment to accommodate the RTX 5090 GPU, and added 3D visualization capabilities. This pr also changes the training to accept three z-slices as input to the model, and output the middle segmentation mask. In future pr(s), I will update the framework to perform semantic segmentation and visualize those segmentations.

### CODE REVIEW

File: callbacks/utils/SaveWholeSlices.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        self,
+        image_dataset: torch.utils.data.Dataset,
+        image_dataset_idxs: list[int],
+        image_specs: dict[str, Any],
+        stride,
+        crop_shape,
+        pad_mode="reflect",
+        image_postprocessor: Any = lambda x: x,
+        local_save_path: Optional[pathlib.Path] = None,
+        **kwargs,
```
Comment (MikeLippincott) [inline callbacks/utils/SaveWholeSlices.py]: consider typing and adding docs for what these args and kwargs are

File: callbacks/utils/SaveWholeSlices.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            crop_shape=crop_shape,
+            stride=stride,
+            pad_slices=True,
+        )
+
+        self.epoch = None
+
+    def reduce_dataset_idxs(self, image_dataset: torch.utils.data.Dataset):
+        """
+        For reducing the dataset to only unique indices.
```
Comment (MikeLippincott) [inline callbacks/utils/SaveWholeSlices.py:47]: I am not quite sure what this means?

An index is what is this case?

File: callbacks/utils/SaveWholeSlices.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+        return output[self.original_crop_coords]
+
+    def pad_image(self, input_image: torch.Tensor) -> torch.Tensor:
+        """
+        input_image:
+            Expects image of shape: (Z, H, W)
+            Z -> Number of Z slices
+            H -> Image Height
+            W -> Image Width
```
Comment (MikeLippincott) [inline callbacks/utils/SaveWholeSlices.py:110]: consider using X,Y,Z as the dimensions - height could be used as z.

File: callbacks/utils/SaveWholeSlices.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                raise ValueError("Pixels should be between 0 and 1 in the input image")
+
+        if image_type == "target":
+            image = (image != 0).float()
+
+        image = (image * 255).byte().cpu().numpy()
+
+        # Black images will not be saved
+        if np.max(image) == 0:
+            return False
```
Comment (MikeLippincott) [inline callbacks/utils/SaveWholeSlices.py:147]: Why not? A black image is still important yes? The absence of singal is still a signal.

File: callbacks/utils/SaveWholeSlices.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+                image_path=self.image_dataset[sample_idx]["target_path"],
+                image_type="target",
+                image=self.image_dataset[sample_idx]["target"],
+            )
+
+            # Only save these images if the segmentation mask isn't black
+            if sample_image:
+                padded_image = self.pad_image(
+                    input_image=self.image_dataset[sample_idx]["input"]
+                )
```
Comment (MikeLippincott) [inline callbacks/utils/SaveWholeSlices.py:216]: Same here - I think a really convincing "why" statement is needed

File: callbacks/utils/image_padding_specs.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
+import pathlib
+from typing import Any, Optional
+
+import numpy as np
+import tifffile
+
+
+def compute_pad(image_dim_size, crop_dim_size, stride):
```
Comment (MikeLippincott) [inline callbacks/utils/image_padding_specs.py]: consider typing and more docs for each function. 

What is stride in this case?

File: callbacks/utils/image_padding_specs.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    Parameters
+    ----------
+    image_specs : dict[str, Any]
+        Dictionary containing image dimensions with the key:
+        - "image_shape": tuple[int]
+            Expects Image shape to be: (Z, H, W)
+            Z -> Number of Z slices
+            H -> Image Height
+            W -> Image Width
```
Comment (MikeLippincott) [inline callbacks/utils/image_padding_specs.py:43]: same comment about using Z,Y,X

File: callbacks/utils/image_padding_specs.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    image_specs: dict[str, Any],
+    crop_shape: tuple[int],
+    stride: tuple[int],
+    pad_slices: bool = True,
+) -> tuple[tuple[tuple[int, int], ...], tuple[slice, ...]]:
+    """
+    Compute the reflection padding required to evenly tile a 3D image
+    with patches of the given crop size and stride. Also returns the
+    bounding box coordinates to recover the original image region
+    after padding.
```
Comment (MikeLippincott) [inline callbacks/utils/image_padding_specs.py:33]: Is reflection padding the best approach here? Might there be a different appraoch??

File: datasets/dataset_00/utils/ImageSelector.py [hunk 9]
```diff
@@ -14,38 +14,75 @@ class ImageSelector:
...
         device: Union[str, torch.device] = "cuda",
     ):
-        self.number_of_slices = number_of_slices
+        """
+        input_crop_shape:
+            Expects the crop shape to be: (Z, H, W)
+            Z -> Number of Z slices
+            H -> Image Height
+            W -> Image Width
+
```
Comment (MikeLippincott) [inline datasets/dataset_00/utils/ImageSelector.py:29]: same here (Z,Y,X)

File: visualize_results_nviz/nbconverted/convert_images.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            thresholded_segmentations_save_path
+            / f"3D_generated-segmentation_{zslice_idx}.tiff"
+        )
+
+        # Save 2D thresholded segmentations images
+        if not segmentation_image_path.exists():
+            tifffile.imwrite(
+                segmentation_image_path,
+                (image[zslice_idx] >= 55).astype(np.uint8) * 255,
+            )
```
Comment (MikeLippincott) [inline visualize_results_nviz/nbconverted/convert_images.py]: It is unclear to me why the images need to saved into a 2D format for nviz? nviz takes 3D images as default. 

For comments like these where design choices were unclear, please consider having a rationale readme doc that explains some of these choices if the documentation should not be in the code. 

Some workflow figures will ultimatly help others to understand these too. More than likely out of scope of this PR but would be helpful for future work! :)

File: visualize_results_nviz/view_images.sh [hunk 11]
```diff
@@ -3,12 +3,12 @@
...
-# Path to the folder containing the mlflow images
+# Path to the folder containing the images
 mlflow_artifacts_path=$1
 
 temp_image_dir=$(mktemp -d)
 
-uv run python convert_images.py "$mlflow_artifacts_path" "$temp_image_dir"
+uv run python nbconverted/convert_images.py "$mlflow_artifacts_path" "$temp_image_dir"
 
 while true; do
```
Comment (MikeLippincott) [inline visualize_results_nviz/view_images.sh]: If you want to generate mp4 or gifs I highly reccomend interacting with nviz in a python script or notebook as nviz returns a `viewer` object that can be interacted with via python libraries!

File: visualize_results_nviz/nbconverted/convert_images.py (hunk 10 above)
Comment (MikeLippincott) [reply visualize_results_nviz/nbconverted/convert_images.py]: My guess is that this is where an error is being introduced in your visualization pipeline. Seems like there is a duplication and spacing of z-slices or something weird like that.

**Review by MikeLippincott — APPROVED** [review]
LGTM. I think a little more work might be needed in documentation external of the code to explain the whole process, training approach, etc.

File: callbacks/utils/SaveWholeSlices.py (hunk 1 above)
Comment (MattsonCam) [reply callbacks/utils/SaveWholeSlices.py]: Good catches. The kwargs isn't needed here

File: callbacks/utils/SaveWholeSlices.py (hunk 2 above)
Comment (MattsonCam) [reply callbacks/utils/SaveWholeSlices.py:47]: > I am not quite sure what this means?

It's complicated, essentially each sample used in the model is a subset of a whole image (crop of an FOV for a given number of slices). If I were to save the whole image (with all slices) for each of these samples (represented as indices), then I would be saving the same image multiple times (because there are multiple (crops in an image).

> An index is what is this case?

Each index is a sample from the dataset used to train the model and save crops, but they become redundant at the whole image level

File: callbacks/utils/SaveWholeSlices.py (hunk 4 above)
Comment (MattsonCam) [reply callbacks/utils/SaveWholeSlices.py:147]: By itself I don't think a black image is very useful. However, if there were cells in the brightfield image and the output segementation image was black, then that could be useful.

For now I think I will keep it as is because the model is only predicting segmentations based on a crop that may or may not contain cells, so some of these crop segmentation will also give completely black segmentations (the difference being that the whole segmented image won't be completely black)

File: callbacks/utils/SaveWholeSlices.py (hunk 5 above)
Comment (MattsonCam) [reply callbacks/utils/SaveWholeSlices.py:216]: Added

File: callbacks/utils/image_padding_specs.py (hunk 6 above)
Comment (MattsonCam) [reply callbacks/utils/image_padding_specs.py]: Good catch.

Stride is the same as in DL frameworks such as keras and pytorch

File: callbacks/utils/SaveWholeSlices.py (hunk 3 above)
Comment (MattsonCam) [reply callbacks/utils/SaveWholeSlices.py:110]: Ya, it depends on how you are thinking of the coordinate system.

File: callbacks/utils/image_padding_specs.py (hunk 7 above)
Comment (MattsonCam) [reply callbacks/utils/image_padding_specs.py:43]: See above response

File: datasets/dataset_00/utils/ImageSelector.py (hunk 9 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImageSelector.py:29]: See above response

File: callbacks/utils/image_padding_specs.py (hunk 8 above)
Comment (MattsonCam) [reply callbacks/utils/image_padding_specs.py:33]: There's always a different approach. Most papers that use padding during inference in biomedical imaging tend to use reflection / mirror padding. I think sometimes those terms are used interchangeably though

Comment (MattsonCam) [reply callbacks/utils/image_padding_specs.py:33]: Right now I'm just including reflection, and if I want to change it later I can

File: visualize_results_nviz/nbconverted/convert_images.py (hunk 10 above)
Comment (MattsonCam) [reply visualize_results_nviz/nbconverted/convert_images.py]: > It is unclear to me why the images need to saved into a 2D format for nviz? nviz takes 3D images as default.

From what I understand there is an [open pr ](https://github.com/WayScience/nViz/pull/64) so that nviz can take in 3D images. I get an error when I try to convert 3D tiff images to an OME-zarr or OME-tiff format

Comment (MattsonCam) [reply visualize_results_nviz/nbconverted/convert_images.py]: > My guess is that this is where an error is being introduced in your visualization pipeline. Seems like there is a duplication and spacing of z-slices or something weird like that.

I'm not seeing any mistakes in the code at this moment. The `visualize_results_nviz` folder is fairly minimal, and the z-slice spacing specified in the bash script looks correct to me.

Comment (MattsonCam) [reply visualize_results_nviz/nbconverted/convert_images.py]: > For comments like these where design choices were unclear, please consider having a rationale readme doc that explains some of these choices if the documentation should not be in the code.
> 
> Some workflow figures will ultimatly help others to understand these too. More than likely out of scope of this PR but would be helpful for future work! :)

Right now I'm in the process of rapidly trying different modeling approaches so I can make time for other priorities while my computer is running. I'll definitely be adding more to the README later

File: visualize_results_nviz/view_images.sh (hunk 11 above)
Comment (MattsonCam) [reply visualize_results_nviz/view_images.sh]: Interesting, I might try this in the future. Napari seems to be good enough for me, and it also allows you to generate videos.

File: callbacks/utils/SaveWholeSlices.py (hunk 3 above)
Comment (MikeLippincott) [reply callbacks/utils/SaveWholeSlices.py:110]: My apologies, this was less of a suggestion and more of a requirement. The standard in microscopy is using XYZ. Any other use is confusing and other biologist *will* be confused.

**Comment by MattsonCam** [comment]
I'm going to wait until we understand the napari/nViz visualization problem better before I merge this pr. Just in case I need to change my code

**Comment by MattsonCam** [comment]
> I'm going to wait until we understand the napari/nViz visualization problem better before I merge this pr. Just in case I need to change my code

After some discussion, I think this problem has be resolved enough to merge the pr, so I am merging it now. Thanks for the review @MikeLippincott!
